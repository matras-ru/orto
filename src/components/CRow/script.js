import { selfInstall } from '@/utils';
import { mergeData } from 'vue-functional-data-merge';

import ThemeClass from '@/themes/default/CRow';

const {
    baseClass,
    smallSizeGuttersNormalizeClass,
    mediumGuttersNormalizeClass,
    smallSizeGuttersClass,
    mediumGuttersClass
} = ThemeClass;

const VALID_GUTTERS = ['none', 'md', 'sm'];
const VALID_DIRECTION = ['row', 'column'];
const VALID_ALIGN = ['start', 'center', 'end'];
const VALID_JUSTIFY = ['start', 'center', 'between', 'end'];

const props = {
    justify: {
        type: String,
        default: null,
        validator: value => VALID_JUSTIFY.includes(value)
    },

    align: {
        type: String,
        default: null,
        validator: value => VALID_ALIGN.includes(value)
    },

    direction: {
        type: String,
        default: 'row',
        validator: value => VALID_DIRECTION.includes(value)
    },

    gutters: {
        type: String,
        default: 'md',
        validator: value => VALID_GUTTERS.includes(value)
    }
};

const currentClass = ({ gutters }) => {
    const rowClasses = [...baseClass];
    const colClasses = [];

    // negative margin div.-mx-4/8/none + positive padding for children cols
    switch (gutters) {
        case 'none':
            break;
        case 'sm':
            rowClasses.push(...smallSizeGuttersNormalizeClass);
            colClasses.push(...smallSizeGuttersClass);
            break;
        case 'md':
        default:
            rowClasses.push(...mediumGuttersNormalizeClass);
            colClasses.push(...mediumGuttersClass);
            break;
    }

    return {
        rowClasses,
        colClasses
    };
};

export default {
    name: 'CRow',

    functional: true,

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    props,

    render(h, { props, data, children = [] }) {
        const { rowClasses, colClasses } = currentClass(props);

        const componentData = {
            class: rowClasses
        };

        const computedChildren = children.map(item => {
            item.data.staticClass = colClasses.join(' '); // add gutters for Cols
            return item;
        });

        return h('div', mergeData(data, componentData), computedChildren);
    }
};
