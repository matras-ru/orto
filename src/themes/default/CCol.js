export const baseClass = '';
export const defaultClass = 'flex-1 max-w-full';

export default {
    baseClass,
    defaultClass
};
