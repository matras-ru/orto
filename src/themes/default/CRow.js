export const baseClass = 'flex flex-wrap';
//
export const lgGuttersNormalizeClass = '-mx-1-4';
export const lgGuttersClass = 'px-1-4';
//
export const mdGuttersNormalizeClass = '-mx-0-8';
export const mdGuttersClass = 'px-0-8';
//
export const smGuttersNormalizeClass = '-mx-0-4';
export const smGuttersClass = 'px-0-4';
//
export const noneGuttersNormalizeClass = '';
export const noneGuttersClass = '';
//
export const betweenClass = 'justify-between';
export const startClass = 'justify-start';
export const endClass = 'justify-end';
//
export const alignClass = '';
// ...

export default {
    baseClass,
    //
    lgGuttersNormalizeClass,
    mdGuttersNormalizeClass,
    smGuttersNormalizeClass,
    noneGuttersNormalizeClass,

    //
    smGuttersClass,
    mdGuttersClass,
    lgGuttersClass,
    noneGuttersClass,
    //
    betweenClass,
    startClass,
    endClass,
    alignClass
};
