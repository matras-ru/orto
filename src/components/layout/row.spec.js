/* eslint-env jest */
import { mount, createLocalVue } from '@vue/test-utils';

import CRow from './row';
import CCol from './col';

const smBreakpoint = 'sm';

const baseClass = 'flex flex-wrap';
const mdGuttersNormalizeClass = '-mx-0-8';
const smGuttersNormalizeClass = '-mx-0-4';

const colBaseClass = 'flex-1 max-w-full';
const mdGuttersClass = 'px-0-8';
const smGuttersClass = 'px-0-4';

// const mdGuttersClass = 'px-0-8';

describe('CRow', () => {
    it('default empty CRow is the functional component and rendered', () => {
        const wrapper = mount(CRow);
        expect(wrapper.isFunctionalComponent).toBe(true);
        expect(wrapper.is('div')).toBe(true);
        expect(wrapper.classes().sort()).toEqual(
            `${baseClass} ${mdGuttersNormalizeClass}`.split(' ').sort()
        );
        expect(wrapper.classes().length).toBe(3);
        expect(wrapper.text()).toEqual('');
    });
});

describe('CRow - gutters', () => {
    const localVue = new createLocalVue();

    it('add CCol gutter class', () => {
        const App = localVue.extend({
            render(h) {
                return h('section', {}, [
                    h(CRow, { ref: 'row-test' }, [h(CCol, { ref: 'col-test' }, 'Col')])
                ]);
            }
        });

        const wrapper = mount(App, {
            localVue
        });

        const Row = wrapper.find({ ref: 'row-test' });
        const Col = wrapper.find({ ref: 'col-test' });

        expect(Row.classes().sort()).toEqual(
            `${baseClass} ${mdGuttersNormalizeClass}`.split(' ').sort()
        );
        expect(Row.classes().length).toBe(3);

        expect(Col.classes().sort()).toEqual(`${colBaseClass} ${mdGuttersClass}`.split(' ').sort());
        expect(Col.classes().length).toBe(3);

        wrapper.destroy();
    });

    it('responsive gutters', () => {
        const App = localVue.extend({
            render(h) {
                return h('div', {}, [
                    h(
                        CRow,
                        {
                            ref: 'row-test',
                            props: {
                                gutters: 'none',
                                guttersSm: 'sm'
                            }
                        },
                        [h(CCol, { ref: 'col-test' }, 'Col')]
                    )
                ]);
            }
        });

        const wrapper = mount(App, {
            localVue
        });

        const Row = wrapper.find({ ref: 'row-test' });
        const Col = wrapper.find({ ref: 'col-test' });

        expect(Row.classes().sort()).toEqual(
            `${baseClass} ${smBreakpoint}:${smGuttersNormalizeClass}`.split(' ').sort()
        );
        expect(Row.classes().length).toBe(3);

        expect(Col.classes().sort()).toEqual(
            `${colBaseClass} ${smBreakpoint}:${smGuttersClass}`.split(' ').sort()
        );
        expect(Col.classes().length).toBe(3);

        wrapper.destroy();
    });
});
