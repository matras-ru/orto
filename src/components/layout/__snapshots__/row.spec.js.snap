// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CRow default 1`] = `<div class="flex flex-wrap -mx-0-8"></div>`;

exports[`cols basic 1`] = `
<div>
  <div class="flex flex-wrap -mx-0-8">
    <div class="max-w-full px-0-8 w-full md:w-1/5">Col</div>
  </div>
</div>
`;

exports[`gutters add CCol gutter class 1`] = `
<section>
  <div class="flex flex-wrap -mx-0-8">
    <div class="max-w-full px-0-8">Col</div>
  </div>
</section>
`;
